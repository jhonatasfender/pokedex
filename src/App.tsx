import { memo } from "react";

import Form from "./components/Form";
import Header from "./components/Header";
import Item from "./components/Item";
import usePokemon from "./hooks/use-pokemon";
import * as S from "./styles";

function App() {
  const { list, favorite, search, listTypes } = usePokemon();

  return (
    <>
      <Header />

      <Form search={search} listTypes={listTypes} />

      <S.Wrapper>
        {list.map((item) => (
          <Item item={item} key={item.national_number} onLike={favorite} />
        ))}
      </S.Wrapper>
    </>
  );
}

export default memo(App);
