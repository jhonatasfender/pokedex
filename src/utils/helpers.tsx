import React from "react";
import { ThemeProvider } from "styled-components";

import { render, RenderResult } from "@testing-library/react";

import theme from "../style-global/theme";

export const renderWithTheme = (children: React.ReactNode): RenderResult =>
  render(<ThemeProvider theme={theme}>{children}</ThemeProvider>);
