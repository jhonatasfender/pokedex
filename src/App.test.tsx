import { logRoles, screen, waitFor } from "@testing-library/react";
import { act } from "@testing-library/react-hooks";
import userEvent from "@testing-library/user-event";
import fetch from "jest-mock-fetch";

import App from "./App";
import { renderWithTheme } from "./utils/helpers";

const Select = (props: any) => {
  const handleOnChange = (e: React.ChangeEvent<HTMLSelectElement>) => {
    if (!props.isMulti) {
      props.onChange({ value: e.target.value });
    } else {
      props.onChange([{ value: e.target.value }]);
    }
  };

  return (
    <select {...props} multiple={!!props.isMulti} onChange={handleOnChange}>
      {props.options.map((p: any) => (
        <option value={p.value} key={p.value}>
          {p.label}
        </option>
      ))}
    </select>
  );
};

jest.mock("react-select", () => ({
  __esModule: true,
  default: Select,
}));

const isDefault = {
  results: [
    {
      national_number: "001",
      evolution: null,
      sprites: {
        normal:
          "https://img.pokemondb.net/sprites/omega-ruby-alpha-sapphire/dex/normal/bulbasaur.png",
        large: "https://img.pokemondb.net/artwork/bulbasaur.jpg",
        animated:
          "https://img.pokemondb.net/sprites/black-white/anim/normal/bulbasaur.gif",
      },
      name: "Bulbasaur",
      type: ["Grass", "Poison"],
      total: 318,
      hp: 45,
      attack: 49,
      defense: 49,
      sp_atk: 65,
      sp_def: 65,
      speed: 45,
    },
    {
      national_number: "002",
      evolution: null,
      sprites: {
        normal:
          "https://img.pokemondb.net/sprites/omega-ruby-alpha-sapphire/dex/normal/ivysaur.png",
        large: "https://img.pokemondb.net/artwork/ivysaur.jpg",
        animated:
          "https://img.pokemondb.net/sprites/black-white/anim/normal/ivysaur.gif",
      },
      name: "Ivysaur",
      type: ["Grass", "Poison"],
      total: 405,
      hp: 60,
      attack: 62,
      defense: 63,
      sp_atk: 80,
      sp_def: 80,
      speed: 60,
    },
  ],
};

describe("App", () => {
  afterEach(() => {
    fetch.reset();
  });

  it("renders learn react link", async () => {
    fetch.mockResolvedValue({
      json: async () => isDefault,
    });

    renderWithTheme(<App />);

    await waitFor(() => {
      expect(screen.queryByRole("img")).toBeInTheDocument();
    });

    expect(document.body).toMatchSnapshot();
  });

  it("renders learn react link1", async () => {
    fetch.mockResolvedValue({
      json: async () => isDefault,
    });

    renderWithTheme(<App />);

    const [select] = screen.getAllByRole("combobox");

    userEvent.selectOptions(select, ["national_number,desc"]);

    await waitFor(() => {
      const [card] = screen.getAllByRole("card");

      expect(card.querySelector("span")?.textContent).toBe("002");
    });

    userEvent.selectOptions(select, ["national_number,asc"]);

    await waitFor(() => {
      const [card] = screen.getAllByRole("card");

      expect(card.querySelector("span")?.textContent).toBe("001");
    });
  });

  it("renders learn react link5", async () => {
    fetch.mockResolvedValue({
      json: async () => isDefault,
    });

    renderWithTheme(<App />);

    const [select] = screen.getAllByRole("combobox");

    userEvent.selectOptions(select, ["national_number,desc"]);

    await waitFor(() => {
      const [card] = screen.getAllByRole("card");

      expect(card.querySelector("span")?.textContent).toBe("002");
    });

    userEvent.type(screen.getByRole("textbox"), "001");

    await waitFor(() => {
      const [card] = screen.getAllByRole("card");

      expect(card.querySelector("span")?.textContent).toBe("001");
    });
  });

  it("renders learn react link2", async () => {
    fetch.mockResolvedValue({
      json: async () => isDefault,
    });

    renderWithTheme(<App />);

    userEvent.click(screen.getByRole("checkbox"));

    await waitFor(() => {
      expect(screen.queryByRole("card")).not.toBeInTheDocument();
    });

    userEvent.click(screen.getByRole("checkbox"));

    await waitFor(() => {
      const [card] = screen.getAllByRole("card");

      expect(card.querySelector("span")?.textContent).toBe("001");
    });

    const [select] = screen.getAllByRole("combobox");

    userEvent.selectOptions(select, ["national_number,desc"]);

    await waitFor(() => {
      const [card] = screen.getAllByRole("card");

      expect(card.querySelector("span")?.textContent).toBe("002");
    });

    const [, , img] = screen.getAllByRole("img");
    userEvent.hover(img);
    userEvent.click(img);
    
    await waitFor(() => {
      const [card] = screen.getAllByRole("card");

      expect(card.querySelector("span")?.textContent).toBe("002");
    });

    userEvent.click(screen.getByRole("checkbox"));

    await waitFor(() => {
      const [card] = screen.getAllByRole("card");

      expect(card.querySelector("span")?.textContent).toBe("002");
    });
  });

  it("renders learn react link3", async () => {
    fetch.mockResolvedValue({
      json: async () => isDefault,
    });

    renderWithTheme(<App />);

    const [select] = screen.getAllByRole("listbox");

    await waitFor(() => {
      expect(select.querySelectorAll("option")).toHaveLength(2);
    });

    userEvent.selectOptions(select, ["Grass"]);

    await waitFor(() => {
      const [card] = screen.getAllByRole("card");

      expect(card.querySelector("span")?.textContent).toBe("001");
    });
  });
});
