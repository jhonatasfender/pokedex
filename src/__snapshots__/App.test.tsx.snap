// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders learn react link 1`] = `
.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  gap: 1rem;
  padding: 1rem;
  z-index: 4;
}

.c3 {
  max-width: 30rem;
  min-width: 30rem;
}

.c2 {
  border: 0.1rem solid hsl(0,0%,80%);
  border-radius: 0.4rem;
  height: 3.6rem;
  padding: 0 1rem;
}

.c2:focus {
  outline: none;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 1rem;
}

.c0 {
  padding: 1rem;
}

.c0 img {
  width: 17rem;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  gap: 2rem;
  padding: 1rem;
  overflow-x: auto;
}

.c5::-webkit-scrollbar {
  width: 0.5rem;
  background-color: #FFFFFF;
}

.c5::-webkit-scrollbar-track {
  border-radius: 1rem;
}

.c5::-webkit-scrollbar-thumb {
  background-color: #3ed3af;
  border-radius: 1rem;
}

.c5::-webkit-scrollbar-thumb:hover {
  background-color: #27f1c0;
}

<body>
  <div>
    <header
      class="c0"
    >
      <img
        alt=""
        src="/img/logo.png"
      />
    </header>
    <form
      class="c1"
      role="form"
    >
      <input
        class="c2"
        name="search"
        type="text"
      />
      <select
        class="c3"
        multiple=""
        options=""
      />
      <select
        options="[object Object],[object Object],[object Object],[object Object]"
      >
        <option
          value="name,asc"
        >
          Nome ASC
        </option>
        <option
          value="name,desc"
        >
          Nome DESC
        </option>
        <option
          value="national_number,asc"
        >
          Registro Nacional ASC
        </option>
        <option
          value="national_number,desc"
        >
          Registro Nacional DESC
        </option>
      </select>
      <div
        class="c4"
      >
        <label
          for="only-like"
        >
          Filtrar somente os favoritados
        </label>
        <input
          id="only-like"
          name="onlyLike"
          type="checkbox"
        />
      </div>
    </form>
    <div
      class="c5"
    />
  </div>
</body>
`;
